### PR TITLE
Add orchard key import/export/production support

### PR DESCRIPTION
**Pull request by AloeareV**
Currently we compile with untested (well, technically there's a test, so insufficiently tested is more accurate) support for importing orchard spending keys. No support yet for doing anything with them, generating them, I don't even think you the code that generates them is user-facing yet.

### CODE REVIEW

File: lib/src/lightwallet.rs [hunk 1]
```diff
@@ -52,6 +52,7 @@ pub(crate) mod keys;
 pub(crate) mod message;
 pub(crate) mod utils;
 pub(crate) mod wallet_transactions;
+mod walletokey;
```
Comment (zancas) [inline lib/src/lightwallet.rs:58]: What a terrible name...

File: lib/src/lightwallet/walletokey.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#[derive(Debug, Clone)]
+pub(crate) enum WalletOKeyInner {
+    HdKey(SpendingKey),
+    ImportedSpendingKey(SpendingKey),
+    ImportedFullViewKey(FullViewingKey),
+    ImportedInViewKey(IncomingViewingKey),
+    ImportedOutViewKey(OutgoingViewingKey),
+}
+
+impl PartialEq for WalletOKeyInner {
```
Comment (zancas) [inline lib/src/lightwallet/walletokey.rs:63]: So we're depending on the name of this enclosing mod `walletokey` to disambiguate among kind of keys...   e.g. `IncomingViewingKey` from `Orchard` as opposed to from `Sapling`?

**Comment by AloeareV** [comment]
As we are capable of now generating, importing, and exporting orchard keys (we can export but not import orchard viewing keys), I'm marking this ready to merge in keeping with the desire for quick incremental changes.

File: lib/Cargo.toml [hunk 3]
```diff
@@ -40,6 +40,7 @@ secp256k1 = "=0.21.3"
 ripemd160 = "0.9.1"
 sha2 = "0.9.5"
 base58 = "0.1.0"
+bech32 = "0.9.0"
```
Comment (zancas) [inline lib/Cargo.toml:43]: Shouldn't an address codec be exported from the address library (`librustzcash`)?

Comment (zancas) [reply lib/Cargo.toml:43]: Issued.

File: lib/Cargo.toml [hunk 4]
```diff
@@ -49,6 +50,9 @@ jubjub = "0.9.0"
 bls12_381 = "0.7"
 group = "0.12"
 rust-embed = { version = "6.3.0", features = ["debug-embed"] }
+orchard = "0.1.0"
```
Comment (zancas) [inline lib/Cargo.toml:53]: It's BOSL.

File: lib/Cargo.toml [hunk 5]
```diff
@@ -49,6 +50,9 @@ jubjub = "0.9.0"
...
+subtle = "2.4.1"
+zcash_address = { git = "https://github.com/zcash/librustzcash", rev = "73d9395c9d3c5fa81fc7becd363a2c1a51772a76"}
```
Comment (zancas) [inline lib/Cargo.toml:55]: Maybe this is UA specific.

File: lib/src/lightclient.rs [hunk 6]
```diff
@@ -273,6 +273,8 @@ impl LightClient {
 
                 Ok(lc)
             })
+        } else if seed_phrase.starts_with(config.chain.hrp_orchard_spending_key()) {
```
Comment (zancas) [inline lib/src/lightclient.rs:276]: What're the tradeoffs implied by this innovation?

File: lib/src/lightclient.rs [hunk 7]
```diff
@@ -369,59 +371,79 @@ impl LightClient {
...
 
         let mut all_keys = vec![];
+        all_keys.extend_from_slice(&o_keys);
         all_keys.extend_from_slice(&z_keys);
         all_keys.extend_from_slice(&t_keys);
 
         Ok(all_keys.into())
     }
 
+    //helper function to export all keys of a type
```
Comment (zancas) [inline lib/src/lightclient.rs:421]: Very nice!

File: lib/src/lightclient.rs [hunk 8]
```diff
@@ -931,18 +954,26 @@ impl LightClient {
...
+                match $key {
+                    $(_ if $key.starts_with($start) => $do,)+
+                    _ => Err(format!(
+                        "'{}' was not recognized as either a spending key or a viewing key",
+                        key,
+                    )),
+                }
+            }
         }
+
```
Comment (zancas) [inline lib/src/lightclient.rs:968]: I will walk over this with `cargo` expand.  I love boiling out ugly repetition.

File: lib/src/lightclient/tests.rs [hunk 9]
```diff
@@ -79,7 +79,7 @@ fn new_wallet_from_phrase() {
 }
 
 #[test]
-fn new_wallet_from_sk() {
+fn new_wallet_from_zsk() {
```
Comment (zancas) [inline lib/src/lightclient/tests.rs:82]: Ahh yes...  the ol' `z` for Sapling convention....

File: lib/src/lightwallet/orchardkeys.rs [hunk 10]
```diff
@@ -0,0 +1,93 @@
+use orchard::keys::{Diversifier, FullViewingKey, IncomingViewingKey, OutgoingViewingKey, Scope, SpendingKey};
+use zcash_address::unified::{Address as UnifiedAddress, Encoding, Receiver, Typecode};
+// A struct that holds orchard private keys or view keys
+#[derive(Clone, Debug, PartialEq)]
+pub struct WalletOKey {
+    pub(super) key: WalletOKeyInner,
+    locked: bool,
+    pub(super) unified_address: UnifiedAddress,
+
+    // If this is a HD key, what is the key number
```
Comment (zancas) [inline lib/src/lightwallet/orchardkeys.rs:10]: ?

File: lib/src/lightwallet.rs [hunk 11]
```diff
@@ -552,6 +556,77 @@ impl LightWallet {
...
+    pub async fn add_imported_ok(&self, osk: String, birthday: u64) -> String {
+        if self.keys.read().await.encrypted {
+            return "Error: Can't import spending key while wallet is encrypted".to_string();
+        }
+
+        use bech32::FromBase32;
+        let key_bytes = match bech32::decode(&osk) {
+            Ok((hrp, bytes, variant)) => {
+                if hrp != self.config.chain.hrp_orchard_spending_key() {
+                    return format!(
```
Comment (zancas) [inline lib/src/lightwallet.rs:570]: Shouldn't this be an `Err`?

File: lib/src/lightwallet.rs [hunk 12]
```diff
@@ -552,6 +556,77 @@ impl LightWallet {
...
+                        "invalid human-readable-part {hrp}, expected {}.",
+                        self.config.chain.hrp_orchard_spending_key()
+                    );
+                }
+                if variant != bech32::Variant::Bech32m {
```
Comment (zancas) [inline lib/src/lightwallet.rs:575]: Also seems like an error...  but...  can it actually occur?  Maybe I will add a test with a legacy bech32 encoding!

File: lib/src/lightwallet.rs [hunk 13]
```diff
@@ -552,6 +556,77 @@ impl LightWallet {
...
+            Err(e) => return format!("{e}"),
+        };
+        // First, try to interpret the key
+        let spending_key: OrchardSpendingKey = match OrchardSpendingKey::from_bytes(key_bytes).into() {
+            Some(k) => k,
+            None => return format!("Error importing spending key"),
+        };
+
+        // Make sure the key doesn't already exist
+        if self
```
Comment (zancas) [inline lib/src/lightwallet.rs:593]: I like this defensive style.  I wonder what inspired this particular check?

File: lib/src/lightwallet.rs [hunk 14]
```diff
@@ -552,6 +556,77 @@ impl LightWallet {
...
+        }
+
+        let extfvk = OrchardFullViewingKey::from(&spending_key);
+        let unified_address = {
+            let okeys = &mut self.keys.write().await.okeys;
+            let maybe_existing_okey = okeys
+                .iter_mut()
+                .find(|wk| wk.key.full_viewing_key() == Some(extfvk.clone()));
+
+            // If the viewing key exists, and is now being upgraded to the spending key, replace it in-place
```
Comment (zancas) [inline lib/src/lightwallet.rs:612]: Wha....  ???

**Review by zancas — APPROVED** [review]
This is good enough to go into `dev`.  We need to add significant testing before releasing.

**Comment by zancas** [comment]
Superceded by:  https://github.com/zingolabs/zecwalletlitelib/pull/74

File: lib/src/lightwallet.rs (hunk 11 above)
Comment (AloeareV) [reply lib/src/lightwallet.rs:570]: Yes. This would necessitate changes to the public API.

**Pull request closed without merging**